### DISCUSSION TITLE
netbox.netbox.nb_lookup - api_filter not found any roles

### DISCUSSION THREAD

**Discussion by dshleg** [netbox-community/ansible_modules#1341] [General]
Hi all! 
Ansible version: 2.17
Netbox version 4.1.4

i trying use nb_lookup for find free outlets on Power distributor. Initially i want find  devices with power distributor role in the rack, but get error not found:
`An unhandled exception occurred while running the lookup plugin 'netbox.netbox.nb_lookup'. Error was a <class 'pynetbox.core.query.RequestError'>, original message: The request failed with code 400 Bad Request: {'role': ['Select a valid choice. Power distributor is not one of the available choices.']}. The request failed with code 400 Bad Request: {'role': ['Select a valid choice. Power distributor is not one of the available choices.']}
`
Task example:
`- name: Obtain list of devices from NetBox
  debug:
      msg: >
        {{item}}
  loop: "{{ query('netbox.netbox.nb_lookup', 'devices',
    api_endpoint='http://<netbox_IP>:8000',
    token='my_token', api_filter='role=\"Power distributor\"') }}"
`
next i try for get all roles:

and receive all roles including "Power distributor"
Task:
`- name: Obtain list of all roles from NetBox
  debug:
      msg: >
        {{item}}
  loop: "{{ query('netbox.netbox.nb_lookup', 'device-roles',
    api_endpoint='http://<netbox_IP>:8000',
    token='my_token') }}"
`
RESULT:
`MSG:

{'key': 5, 'value': {'id': 5, 'url': 'http://<netbox_IP>:8000/api/dcim/device-roles/5/', 'display_url': 'http://<netbox_IP>:8000/dcim/device-roles/5/', 'display': 'Power distributor', 'name': 'Power distributor', 'slug': 'power-distributor', 'color': 'ff5722', 'vm_role': False, 'config_template': None, 'description': '', 'tags': [], 'custom_fields': {}, 'created': '2022-06-16T19:34:11.923602Z', 'last_updated': '2024-10-18T22:06:13.076080Z', 'device_count': 12, 'virtualmachine_count': 0}}

ok: [shv01] => (item={'key': 11, 'value': {'id': 11, 'url': 'http://<netbox_IP>:8000/api/dcim/device-roles/11/', 'display_url': 'http://<netbox_ip>:8000/dcim/device-roles/11/', 'display': 'RDP Terminal Server', 'name': 'RDP Terminal Server', 'slug': 'rdp-terminal-server', 'color': '795548', 'vm_role': True, 'config_template': None, 'description': '', 'tags': [], '`

what wrong?

**Comment by boomsfib** [comment]
Change your api_filter to look for the slug.  So...

"{{ query('netbox.netbox.nb_lookup', 'devices', api_endpoint='http://<netbox_IP>:8000', token='my_token', api_filter='role=power-distributor') }}"

**Reply by dshleg** [reply]
thx for reply, but this not work

**Reply by boomsfib** [reply]
What is the error you are receiving now?  Can you post your updated query?  I am actively working on a ansible playbook now that is pulling devices based on roles and this is the query I am using and it works fine for me.

**Reply by dshleg** [reply]
You example:
`"{{ query('netbox.netbox.nb_lookup', 'devices', api_endpoint='http://<netbox_IP>:8000', token='my_token', api_filter='role=power-distributor') }}"`
Not work. 

also tried:
`"{{` query('netbox.netbox.nb_lookup', 'devices', api_endpoint='http://<netbox_IP>:8000', token='my_token', api_filter='slug=power-distributor') `}}"`
my guess- this is because api returns in answer  role field contains nested dict - not string!

Example:
`{'key': 283, 'value': {'id': 283, 'url': 'http://netbox_ip:8000/api/dcim/devices/283/', 'display_url': 'http://netbox_ip:8000/dcim/devices/283/', 'display': '4H1D010-PDU-A', 'name': '4H1D010-PDU-A', 'device_type': {'id': 7, 'url': 'http://netbox_ip:8000/api/dcim/device-types/7/', 'display': 'АР8881', 'manufacturer': {'id': 7, 'url': 'http://netbox_ip:8000/api/dcim/manufacturers/7/', 'display': 'APC', 'name': 'APC', 'slug': 'apc', 'description': ''}, 'model': 'АР8881', 'slug': 'AP8881', 'description': ''}, 'role': {'id': 5, 'url': 'http://netbox_ip:8000/api/dcim/device-roles/5/', 'display': 'Power distributor', 'name': 'Power distributor', 'slug': 'power-distributor', 'description': ''}, 'tenant': {'id': 3, 'url': 'http://netbox_ip:8000/api/tenancy/tenants/3/', 'display': 'ITER', 'name': 'ITER', 'slug': 'iter', 'description': ''}, 'platform': None, 'serial': '', 'asset_tag': None, 'site': {'id': 1, 'url': 'http://netbox_ip:8000/api/dcim/sites/1/', 'display': 'DTLN', 'name': 'DTLN', 'slug': 'dtln', 'description': ''}, 'location': {'id': 1, 'url': 'http://netbox_ip:8000/api/dcim/locations/1/', 'display': 'NORD2', 'name': 'NORD2', 'slug': 'nord2', 'description': '', 'rack_count': 0, '_depth': 0}, 'rack': {'id': 1, 'url': 'http://netbox_ip:8000/api/dcim/racks/1/', 'display': '4H1D010 (4H1D010)', 'name': '4H1D010', 'description': ''}, 'position': None, 'face': None, 'latitude': None, 'longitude': None, 'parent_device': None, 'status': {'value': 'active', 'label': 'Active'}, 'airflow': None, 'primary_ip': None, 'primary_ip4': None, 'primary_ip6': None, 'oob_ip': None, 'cluster': None, 'virtual_chassis': None, 'vc_position': None, 'vc_priority': None, 'description': '', 'comments': '', 'config_template': None, 'config_context': {}, 'local_context_data': None, 'tags': [], 'custom_fields': {}, 'created': '2024-10-18T11:26:12.205790Z', 'last_updated': '2024-10-18T11:26:12.205809Z', 'console_port_count': 0, 'console_server_port_count': 0, 'power_port_count': 1, 'power_outlet_count': 42, 'interface_count': 1, 'front_port_count': 0, 'rear_port_count': 0, 'device_bay_count': 0, 'module_bay_count': 0, 'inventory_item_count': 0}}`

This working for me:
`- name: Obtain list of devices from NetBox -test 2
  debug:
      msg: "{{item | community.general.json_query('role[?slug==\"power-distibutor\"]') }}"
  loop: "{{ query('netbox.netbox.nb_lookup', 'devices',
    api_endpoint='http://<netbox_IP>:8000', token='my_token', api_filter='name__ic=4H1D010-PDU') }}"
`
